sdks/python/apache_beam/coders/row_coder.py

```diff
@@ -163,10 +165,15 @@ def _nonnull_coder_from_type(field_type):
         _coder_from_type(field_type.map_type.key_type),
         _coder_from_type(field_type.map_type.value_type))
   elif type_info == "logical_type":
-    # Special case for the Any logical type. Just use the default coder for an
-    # unknown Python object.
     if field_type.logical_type.urn == PYTHON_ANY_URN:
+      # Special case for the Any logical type. Just use the default coder for an
+      # unknown Python object.
       return typecoders.registry.get_coder(object)
+    elif field_type.logical_type.urn == DATETIME_URN:
+      # Special case for datetime logical type.
+      # DATETIME_URN explicitly uses TimestampCoder which deals with fix length
+      # 8-bytes big-endian-long instead of varint coder.
+      return TimestampCoder()
```

Comment:
If we do implement `beam:logical_type:datetime` in Python we should:
- Add the logical type in schema.proto, and reference the URN from that file: https://github.com/apache/beam/blob/4799828f75309fd494e14eab5af40559e9a140a3/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/schema.proto#L117
- Consider if we can keep the logic contained in the logical type implementation, e.g. can we do the byte flipping in a to_language_type and to_representation_type implementation?
- Add some test cases to standard_coders.yaml, like this one: https://github.com/apache/beam/blob/4799828f75309fd494e14eab5af40559e9a140a3/model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml#L449-L454 You might also need to skip this case in Go if the "logical" filter doesn't catch it: https://github.com/apache/beam/blob/4799828f75309fd494e14eab5af40559e9a140a3/sdks/go/test/regression/coders/fromyaml/fromyaml.go#L51

Reply:
Thanks for the instructions! Will complete the proto change. `beam:logical_type:datetime:v1` is the millisecond precision timestamp backed by a fixed length INT64 that encoded with BigEndianLong. Will complete the proto change.

> can we do the byte flipping in a to_language_type and to_representation_type implementation?

Considered this before. Under the current framework the value is already decoded with VarInt coder before sent to to_language_type or to_representation_type which is incorrect. All I need is a 8-byte fixed long integer primitive which does not exist in portable primitives, and there is not yet a consensus on reference coder  ([context](https://docs.google.com/document/d/1uu9pJktzT_O3DxGd1-Q2op4nRk4HekIZbzi-0oTAips/edit?disco=AAAADKzJ_Q8)). Therefore I ended up this solution to make things work while not introducing new primitives or other fundamental changes.

Reply:
ah... test case in standard_coders.yaml won't work either because Both micros_instant and millis_instant has Timestamp language type and the former will take over the priority of millis_instant (by design). When a timestamp is decoded to an integer, the test will call micros_instant's to_language_type.

Reply:
Hmm I'd really like to have this tested at the `standard_coders.yaml` level as a unit test, it's not ideal if the only thing verifying compatibility is an integration test in Python.

If you have time tomorrow maybe we could do a quick video call to live debug what's going wrong here, and see if we can work around it. If it will be painful we can leave it as future work, but I'd like to understand the level of effort.

Reply:
Managed to get a test case using the following code snippet

```python

def generate_millis():
    # Logical type that deals with millis_instant urn (MillisInstant)
    MillisLogicalType = LogicalType._known_logical_types.get_logical_type_by_urn('beam:logical_type:millis_instant:v1')
    # Original Logical type used to represent Timestamp (MicrosInstant)
    TimestampLogicalType = LogicalType._known_logical_types.get_logical_type_by_language_type(Timestamp)
    
    LogicalType._known_logical_types.by_language_type[Timestamp] = MillisLogicalType

    schema = beam.typehints.schemas.named_tuple_to_schema(TestTuple)
    coder = beam.coders.row_coder.RowCoder(schema)
    print("payload = %s" % schema.SerializeToString())
    examples = (TestTuple(
        f_timestamp=Timestamp.from_rfc3339("2020-08-13T14:14:14.123Z"),
        f_string="2020-08-13T14:14:14.123Z",
        f_int=1597328054123),)
    for example in examples:
        print("example = %s" % coder.encode(example))
    
    # recover original registration
    LogicalType._known_logical_types.by_language_type[Timestamp] = TimestampLogicalType

```

The workaround is temporarily change the mapping of Timestamp -> MillisInstant logical type. Without it Timestamp always maps to MicrosInstant logical type in Python.

Reply:
Ahhh ok, so it looks like the issue was using Python to generate the test values. Another strategy might be to create the schema proto directly. But this way works too and is probably less code, thanks!